src/controller/controller.ino

```diff
@@ -53,16 +79,204 @@ uint16_t make_id(uint8_t id, bool priority, uint8_t type) {
+  module_info.id = module_id;
+  return module_info;
+}
+
+
+uint8_t check_solved() {
+  uint8_t solved = 1;
+  for (int i; i<32; i++) {
+    if (unsolved_puzzles != 0) {
+      solved = 0;
+      break;
+    }
+  }
+  return solved;
+}
+
+
+void add_module_to_bit_vector(uint8_t module_id) {
+  uint8_t byte_index = module_id >> 3;
+  uint8_t bit_index = module_id & 0x07;
+  unsolved_puzzles[byte_index] |= 0x1 << bit_index;
+}
+
+
+void solve_module_in_bit_vector(uint8_t module_id) {
+    uint8_t byte_index = module_id >> 3;
+    uint8_t bit_index = module_id & 0x07;
+    unsolved_puzzles[byte_index] &= ~(0x1 << bit_index);
+}
+
+
+void send_message(uint8_t* message, uint8_t length) {
+  struct can_frame send_frame;
+  
+  send_frame.can_id = make_id(OBUS_CONTROLLER_ID, false, OBUS_TYPE_CONTROLLER);
+  send_frame.can_dlc = length;
+
+  memcpy(send_frame.data, message, OBUS_MSG_LENGTH);
+ 
+  mcp2515.sendMessage(&send_frame);
+}
+
+
 void start_hello() {
 	state = STATE_HELLO;
+  hello_round_start = millis();
+  nr_connected_modules = 0;
+
+  // Zero bit vectors
+  for (int i; i<32; i++) {
+    unsolved_puzzles[i] = 0;
+  }  
+
+  uint8_t message[OBUS_MSG_LENGTH];
+  message[0] = OBUS_MSGTYPE_C_HELLO;
+
+  send_message(message, 1);
+  
+	Serial.println("Start of discovery round");
+}
+
+
+void send_ack() {
+  uint8_t message[OBUS_MSG_LENGTH];
+  message[0] = OBUS_MSGTYPE_C_ACK;
+
+  send_message(message, 1);
+}
+
+
+void receive_hello() {
+  struct can_frame receive_frame;
+  uint16_t current_time = millis();
+  
+  if (mcp2515.readMessage(&receive_frame) == MCP2515::ERROR_OK) {
+    if (receive_frame.data[0] ==  OBUS_MSGTYPE_M_HELLO) {
+      struct module new_module = get_module_info(receive_frame.can_id);
+      Serial.print("Registered module ");
+      Serial.println(new_module.id);
+      connected_modules_ids[nr_connected_modules] = new_module;
+      nr_connected_modules++;
 
-	char message[OBUS_MSG_LENGTH + 1];
-	message[OBUS_MSG_LENGTH] = '\0';
-	send_frame.can_id = make_id(OBUS_CONTROLLER_ID, false, OBUS_TYPE_CONTROLLER);
-	send_frame.can_dlc = OBUS_MSG_LENGTH;
-	mcp2515.sendMessage(&send_frame);
-	Serial.println("sent");
+      if (new_module.type == OBUS_TYPE_PUZZLE) {
+        add_module_to_bit_vector(new_module.id);    
+      }
+      
+      send_ack();
+      Serial.println("ACK");
+    }
+  } else if (current_time - hello_round_start > OBUS_DISC_DURATION * 1000) {
+    state = STATE_GAME;
+    Serial.println("End of discovery round");
+  }
 }
 
+
+void initialize_game() {
+  strikeouts = 0;
+
+  uint16_t game_duration_millis = (uint16_t) OBUS_GAME_DURATION * 1000;
+
+  uint8_t message[OBUS_MSG_LENGTH];
+  message[0] = OBUS_MSGTYPE_C_GAMESTART;
+  message[1] = (uint8_t) ((game_duration_millis & 0xFF000000) >> 0x18);
+  message[2] = (uint8_t) ((game_duration_millis & 0x00FF0000) >> 0x10);
+  message[3] = (uint8_t) ((game_duration_millis & 0x0000FF00) >> 0x08);
+  message[4] = (uint8_t) (game_duration_millis & 0x000000FF);
```

Comment:
Nice! I like that you didn't assume endianess and directly copied the `uint32_t` into the buffer.